README.md
# aarep

Amazonアソシエイトのレポートを分析する簡単ツール

```
perl aarep.pl -u UNIT -m MODE -d DATE INPUT.csv
```

- INPUT: 対象（入力）ファイル
  - アソシエイトのページからダウンロードできる、注文または売上のCSVファイル。
  - zipファイルとしてダウンロードできるのでそれを解凍、すると拡張子csvのCSVファイルが出てくる。それ。
  - 以下、説明のため、例えば2017年7月分の注文データのCSVファイル15...Fee-Order....csv は FO201707.csv、同じく売上のCSVファイルはFE201707.csv という名前に変えて扱う。
- 出力単位: "-u" で指定。
  - y : 年単位
  - ym : 年月
  - ymd : 年月日
  - ymdh : 年月日時
  - h : 時
  - hm : 時分
  - hms : 時分秒
  - w : 曜日
  - wh : 曜日x時
- モード: "-m" で指定。
  - i : item。ASIN単位で表示
  - t : traking id。XXX-22などの。
  - s : ストア
  - d : デバイス
  - "," でつなぐことができる : i,t i,t,s など
- 日付指定: "--date", "-d" で指定。
  - 日付（年、年月、年月日）の指定を行う。
  - 例: "2017-09", "2017-09-10", "2017-09-(1[5-9]|20)"

## 実行例

出力サンプルには架空のデータを用いています。

売上：2017年7月の商品数と紹介料の合計
```
perl aarep.pl FE201707.csv
all	497	17598
```

売上：2017年7月7日の商品数と紹介料の合計
```
perl aarep.pl --date 2017-07-07 FE201707.csv
all	66	1826
```

注文：2017年7月の商品数と売上高（価格）の合計
```
perl aarep.pl FO201707.csv
all	501	287949
```

売上：2016年の月ごとの商品数と紹介料
```
aarep.pl -u ym FE2016.csv
2016-01	1320	36761
2016-02	244	9473
2016-03	214	8444
2016-04	221	9579
2016-05	245	10051
2016-06	691	10297
2016-07	408	13693
2016-08	774	31069
2016-09	1362	43736
2016-10	2807	97682
2016-11	1351	72919
2016-12	2543	90049
```

売上：2017年7月の日別の商品数と紹介料
```
aarep.pl -u ymd FE201707.csv
2017-07-01	243	7475
2017-07-02	269	8009
2017-07-03	265	8896
2017-07-04	317	8670
2017-07-05	266	6714
2017-07-06	256	6291
2017-07-07	266	8265
2017-07-08	291	8845
```

注文：2017年7月の時間別の注文数と価格合計
```
aarep.pl -u h FO201707.csv
00	325	161568
01	203	113887
02	123	73227
03	78	34152
04	52	26497
05	58	64782
...
21	435	226965
22	454	254015
23	501	258738
```

注文：2017年1-6月の曜日別の注文数と価格合計
```
aarep.pl -u w FO2017H1.csv
Sun	289	165983
Mon	193	96964
Tue	175	85747
Wed	197	102262
Thu	226	125493
Fri	327	188071
Sat	294	168263
```

注文：2017年7月の商品の注文数ランキング（注文数と価格）
```
aarep.pl -m i FO201707.csv
all	B00KYEH7GW	15	1070	メイドインアビス（１） (バンブーコミックス)
all	B06ZZH4CKN	14	1400	海外ドラマはたった350の単語でできている
all	B00O438BF0	14	430	ラーメン大好き小泉さん（１） (バンブーコミックス)
...
```

売上：2017年7月の商品の売上ランキング（売上数と紹介料）
```
aarep.pl -m i FE201707.csv
all	B00KYEH7GW	15	85	メイドインアビス（１） (バンブーコミックス)
all	B06ZZH4CKN	14	100	海外ドラマはたった350の単語でできている
all	B00O438BF0	14	35	ラーメン大好き小泉さん（１） (バンブーコミックス)
...
```

売上：2017年7月のトラッキングID別売上ランキング（売上数と紹介料）
```
aarep.pl -m t FE201707.csv
all	example1-22	205	6992
all	example0-22	142	4872
all	example5-22	85	2901
...
```

売上：2017年7月のストア別売上ランキング（売上数と紹介料）
```
aarep.pl -m s FE201707.csv
all	本	53	1816
all	ドラッグストア・ビューティー	44	2749
all	食品＆飲料	31	3456
...
```

売上：2017年7月のデバイス別売上ランキング（売上数と紹介料）
```
aarep.pl -m d FE201707.csv
all	DESKTOP	262	9616
all	PHONE	174	5895
all	TABLET	61	2086
```

売上：特定のトラッキングIDでの商品売上ランキングを得る
```
aarep.pl -m i,t FE201707.csv | grep example4-22
all	B002RUYBUQ	example4-22	3	150	サクセス ステップカラー 110g
all	B01BWK3SOQ	example4-22	3	75	トクホ 江崎グリコ ポスカ<グレープ>エコパウチ 初期虫歯対策ガム 75g
all	B00Z60G3P0	example4-22	2	16	マルちゃん 麺づくり鶏だし塩 86g
...
```

売上：特定の日付での商品売上ランキングを得る
```
aarep.pl -u ymd -m i FE201707.csv | grep 2017-07-07
2017-07-07	B06XNVS99T	5	120	ちおちゃんの通学路　6 (コミックフラッパー)
2017-07-07	B072N7Q34M	5	385	ひきこもらない (幻冬舎単行本)
2017-07-07	B009KWUN7Y	4	28	行け！稲中卓球部（１） (ヤングマガジンコミックス)
...
```





ChangeLog
2019-07-08  Tatsuwo  <yto at example.com>

	* aarep.pl: エラー対処。
	line_seen のチェックをヘッダ認識部分より後にした。

2017-09-20  Tatsuwo  <yto at example.com>

	* aarep.pl: add command line optiton "--date"
	日付（年、年月、年月日）の指定を行う。
	例：
	--date "2017-09"
	--date "2017-09-10"
	--date "2017-09-(1[5-9]|20)"

2017-09-06  Tatsuwo  <yto at example.com>

	* aarep.pl: add "ymdhm", "ymdhms"

2017-08-15  Tatsuwo  <yto at example.com>

	* conv-aarep-old2new.pl: add

2017-08-13  Tatsuwo  <yto at example.com>

	* start!



conv-aarep-old2new.pl
#!/usr/bin/env perl
# -*- coding: utf-8 -*-
# 2016年よりも前の売上レポートCSVを現行フォーマットに変換する
use strict;
use warnings;
use utf8;
use open ":utf8";
binmode STDIN, ":utf8";
binmode STDOUT, ":utf8";

my @_month_names = qw(Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec);
my %month = map {$_month_names[$_] => $_ + 1} 0..11;

my $sep = "\t";
my @labels;
while (<>) {
    chomp;
    if (/^売上レポート /) {
	print "Fee-Earnings reports from ...\n";
	next;
    } elsif (/^ストア$sep/) {
	@labels = map {s/\(.+?\)$//; s/\s+$//; $_} split($sep, $_);
	print join(",", @labels)."\n";
	next;
    }
    my @c = split($sep, $_);
    next if @c <= 1;
    my %h = map {$labels[$_] => $c[$_]} 0..$#c;

    my ($mn, $d, $y) = split(/[ ,]+/, $h{発送日});
    my $m = sprintf("%02d", $month{substr($mn, 0, 3)});
    $h{発送日} = "$y-$m-$d";
    $h{商品名} =~ s/,/，/g;
    $h{紹介料率} =~ s/\%.*$//;

    print join(",", map {$h{$_}} @labels)."\n";
}



aarep.pl
#!/usr/bin/env perl
# -*- coding: utf-8 -*-
use strict;
use warnings;
use Getopt::Long;
use Time::Local;
use POSIX qw(strftime);
use utf8;
use open ":utf8";
binmode STDIN, ":utf8";
binmode STDOUT, ":utf8";

my $unit_time = "all";
# y:year, ym:month, ymd:day, w:dow, h:hour,
# hm:hour-min, hms:hour-min-sec,
# ymdh, ymdhm, ymdhms, wh, all:all
my $mode = ""; 
# "":count, i:item, t:trackingid, s:store, d:device"
my $date = "";
# 日付（年、年月、年月日）の指定： -date "2017-09-10", -date "2017-09-(1[5-9]|20)"

GetOptions(
    "unit=s" => \$unit_time,
    "mode=s" => \$mode,
    "date=s" => \$date,
    );

my @mode_ids = map {
    /^t/ ? "トラッキングID"
	: /^s/ ? "ストア"
	: /^i/ ? "ASIN"
	: /^d/ ? "デバイスの種類"
	: /^n/ ? "商品リンク経由" # only "o"
	: ""
} split(/,/, $mode);

my $report_type = "e"; # e:ernings, o:orders
my $sep = ",";

my @labels;
my %stat;
my %line_seen;
while (<>) {
    chomp;
    if (/^Fee-(.).+? reports from/) {
	$report_type = lc($1); # "e" or "o"
	next;
    } elsif (/^ストア$sep/) {
	@labels = map {s/\(.+?\)$//; s/\s+$//; $_} split($sep, $_);
	next;
    }
    next if defined $line_seen{$_};
    $line_seen{$_} = 1;
    my @c = split($sep, $_);
    #    next if @c < 11 or $c[4] !~ /-22$/;
    next if @c <= 1;
    my %h = map {$labels[$_] => $c[$_]} 0..$#c;
    my ($y, $m, $d, $H, $M, $S);
    if ($report_type eq "e") {
#	($y, $m, $d) = $h{発送日} =~ /^(\d+)-(\d+)-(\d+)/;
	my $_hms;
	($y, $m, $d, $_hms, $H, $M, $S) = $h{発送日} =~ /^(\d+)-(\d+)-(\d+)( (\d+):(\d+):(\d+))?/;
	$H ||= 0;
	$M ||= 0;
	$S ||= 0;
    } elsif ($report_type eq "o") {
	($y, $m, $d, $H, $M, $S) = $h{日付} =~ /^(\d+)-(\d+)-(\d+) (\d+):(\d+):(\d+)/;
    } else {
	die "bad report type [$report_type]";
    }
    next if $date and "$y-$m-$d" !~ /^$date/;
    my $dow = (localtime timelocal(0,0,0,$d,$m-1,$y))[6];
    my $str = $unit_time eq "ymd" ? "$y-$m-$d"
	: $unit_time eq "ym" ? "$y-$m"
	: $unit_time eq "y" ? "$y"
	: $unit_time eq "w" ? $dow
	: $unit_time eq "wh" ? "$dow-$H"
	: $unit_time eq "h" ? "$H"
	: $unit_time eq "hm" ? "$H:$M"
	: $unit_time eq "hms" ? "$H:$M:$S"
	: $unit_time eq "ymdh" ? "$y-$m-$d $H"
	: $unit_time eq "ymdhm" ? "$y-$m-$d $H:$M"
	: $unit_time eq "ymdhms" ? "$y-$m-$d $H:$M:$S"
	: "all";
    my $keys = join("\t", map {$h{$_}} @mode_ids);
    if ($report_type eq "e") {
	my $num = $h{発送済み商品} - ($h{返品済み商品}||0);
	if (defined $h{売上}) {
	    $stat{$unit_time}{$str}{$keys}{Gross} += $h{売上};
	} else {
	    $stat{$unit_time}{$str}{$keys}{Gross} += $h{価格} * $num;
	}
	$stat{$unit_time}{$str}{$keys}{Income} += $h{紹介料};
	$stat{$unit_time}{$str}{$keys}{Num} += $num;
    } elsif ($report_type eq "o") {
	$stat{$unit_time}{$str}{$keys}{Gross} += $h{価格} * $h{数量};
	$stat{$unit_time}{$str}{$keys}{Num} += $h{数量};
    }
    $stat{$unit_time}{$str}{$keys}{Name} = $h{商品名};
}

foreach my $t (sort keys %{$stat{$unit_time}}) {
    my $r = $stat{$unit_time}{$t};
    my @res;
    foreach my $i (sort keys %$r) {
	next if $r->{$i}{Num} <= 0;
	my @info;
	push @info, $i if $i ne "";
#	push @info, (qw(Sun Mon Tue Wed Thu Fri Sat))[$i] 

	push @info, $r->{$i}{Num};
	push @info, $r->{$i}{Income} if $report_type eq "e";
	push @info, $r->{$i}{Gross};
	if (grep {/^ASIN/} @mode_ids) {
	    push @info, $r->{$i}{Name};
	} else {
#	    my $tanka = $r->{$i}{Num} ? $r->{$i}{Income}/$r->{$i}{Num} : 0;
#	    push @info, int($tanka*100)/100;
	}
	push @res, \@info;
    }
    foreach my $i (sort {$b->[1] <=> $a->[1]} @res) {
	my $l = $unit_time eq "w"
	    ? (qw(Sun Mon Tue Wed Thu Fri Sat))[$t] : $t;
	print join("\t", $l, @$i)."\n";
    }
}
